Content.Shared/Clothing/EntitySystems/ClothingSystem.cs

```diff
@@ -101,7 +101,8 @@ private void ToggleVisualLayers(EntityUid equipee, HashSet<HumanoidVisualLayers>
                 {
                     if (comp.Slots.Contains(layer))
                     {
-                        if (TryComp(item, out ClothingComponent? clothing) && clothing.Slots == slot.SlotFlags)
+                        if (TryComp(item, out ClothingComponent? clothing) && clothing.InSlot != null
+                                                                           && clothing.InSlot.Equals(slot.Name))
                         {
```

Comment:
I don't think this is the right way to fix this, and AFAIK `clothing.InSlot.Equals(slot.Name)` should always be true here? I think this likely should just be a has-flag check?

Though clothing & visuals also seem to generally have shitty documentation, but IIRC `ClothingComponent.Slots` is meant to define the slots in which a piece of clothing is considered "worn" i.e., magboots in pockets are clothing in a slot, but not the right kind of slot.

But if you want even more control about what layers clothing will hide when in certain slots, then `HideLayerClothingComponent` probably either needs to have its own set of slot-flags to decide whether it should be enabled or not, or for even more control the hidden layers should be specified per-slot by slot(`i.e., replace `HashSet<HumanoidVisualLayers>` with a `Dictionary<SlotFlags, HashSet<HumanoidVisualLayers>>`).

In general ToggleVisualLayers() seems pretty cursed / convoluted and hard to parse and should probably be re-written.

Reply:
Nope. I tried HasFlag, but you can also try it if you like. Use `clothing.Slots.HasFlag(slot.SlotFlags)` in that section I replaced instead. Add suitstorage to a mask's slots, play a lizard, equip the mask in your suit storage, and watch your nose disappear despite the mask not being on your face. It's because HasFlag is checking whether the piece _can_ cover that layer, not whether it _is_ covering that layer.

`ClothingComponent.Slots` is the sum of the SlotFlags of all its equippable slots. `ClothingComponent.InSlot` is the name of the slot it is currently equipped in.

So with the example of RMC ([where I discovered this issue](https://github.com/RMC-14/RMC-14/issues/5173)), the gas mask can go onto the suit storage slot (for... CM13 parity reasons I guess). `ClothingComponent.Slots` will always contain Mask and Suit Storage SlotFlags. But `ClothingComponent.InSlot` will only be Mask if it's currently in the Mask slot, and only be SuitStorage when it's currently in the SuitStorage slot.

A cleaner way might be to add a SlotFlag equivalent to `ClothingComponent.InSlot`--like `ClothingComponent.InSlotFlag` or something, so we could just do `clothing.InSlotFlag == slot.SlotFlags`, But that would mean adding the code to update that variable everywhere it needs to be updated.

Reply:
Anyway, `HideLayerClothingComponent` _does_ have a slot-flags in a sense. It has "slots", though that is a misnomer--what it calls slots are actually layers. But those layers have associated slots.

So gas mask's HideLayerClothingComponent has several "slots" (layers) including "snout". Each of those layers has corresponding actual slots (snout has Mask and Head IIRC) and then in this method we're checking the equipped clothing against each of these slots. For clothes that can only be worn in one slot, this is fine. And I can't think of any other examples of clothes that can be worn in multiple slots--it's only the RMC gas mask as far as I know.

Reply:
Okay I take it back. I thought I understood the code, I did not not understand the code. Also somehow I had foolishly tested the change with a default snout rodentia model on RMC (i.e. flat-faced) and so hadn't noticed that my proposed fix didn't even actually fix the issue. And thinking that it had fixed the issue, I _really_ misunderstood the code.

I went back over it with a debugger and you're absolutely right, this is an even bigger mess than I'd realized. HideLayerClothingComponent would need a refactor to fix this issue. I'm not 100% sure how to manage that refactor without breaking compatibility with every yaml definition that uses it though. I'll think on it and see what I can come up with.

Reply:
Yeah like I said:

> In general ToggleVisualLayers() seems pretty cursed / convoluted and hard to parse and should probably be re-written.

If you do just change it to use its own slot flags, you could make them default to using the clothing flags which might help with compatibility? But either way the code probably needs rewriting to actually make it intelligible.

Reply:
Things got delayed for the holidays, but I've made some progress.

As you'll see, I've begun the refactor you recommended and then some. Item prototypes that use HideLayerClothingComponent.Slots still work the same (and will continue to exhibit the bug this is fixing--no way around that). Item prototypes using the new HideLayerClothingComponent.Layers will not exhibit the bug. Item prototypes using both will default to Layers and ignore Slots.

I've also made ToggleVisualLayers much more CPU efficient by avoiding the nested loops. I've discovered why that loop was nested though, as this does introduce a _new_ bug: wearing multiple clothes that hide the same layer (e.g. gas mask + welding mask) can mess with the layer visibility, especially if you then begin toggling the mask on and off.

I have an idea for how to address it--by similarly dressing up HumanoidAppearanceComponent.HiddenLayers into a Dictionary, so that we can track all slots currently hiding a given layer. I didn't want to commit to this before first getting feedback on the approach I've already taken though.